kala.py

```diff
@@ -37,43 +39,185 @@
+        return document
+    staging = os.environ.get('KALA_FILTER_STAGING', app.config['filter.staging'])
+    object_id = mongodb[staging].insert(document)
+    cursor = mongodb[staging].find(filter=filter_json)
+    # Delete from staging collection after cursor becomes a list, otherwise cursor will produce an empty list.
+    documents = [doc for doc in cursor]
+    mongodb[staging].remove({"_id": object_id}, "true")
+    return documents
+
+
+def _filter_read(document):
+    """This is used to filter the JSON object."""
+    # This will throw if setting isn't found in config.
+    # Expected as without setting you have nothing to filter.
+    whitelist = app.config['filter.read']
+    if whitelist is None:
+        return document
+    # When document is a dictionary, deletes any keys which are not in the whitelist.
+    # Unless they are an operator, in which case we apply the filter to the value.
+    if isinstance(document, dict):
+        for key in list(document.keys()):
+            if key.startswith('$'):
+                _filter_read(document[key])
+            elif key not in whitelist:
+                del document[key]
+    # When document is a list, apply the filter on each item, thus returning a filtered list.
+    elif isinstance(document, list):
+        document = [item for item in document if _filter_read(item)]
+    # When document is a tuple, return whether first element is in the whitelist.
+    # This is used for sort
+    elif isinstance(document, tuple):
+        return document[0] in whitelist
+    # This is used for projection.
+    # Note that a JSON object can not contain null values.
+    elif document is None:
+        document = dict((key, '1') for key in whitelist)
+    return document
+
+
+def _filter_aggregate(list_):
+    """This is used to filter the aggregate JSON
+
+    Keyword arguments:
+    list_ -- The JSON should be a list of dictionaries.
+    """
+    # The idea is to insert a $project at the start of pipeline that only contains fields in the whitelist.
+    # Once filtered, the user can do whatever they want and never touch sensitive data.
+    project = {'$project': dict((field, 1) for field in app.config['filter.read'])}
+    list_ = [project] + list_
+    return list_
+
+
+def _convert_object_type(document, type_):
+    """This is used to convert strings to the correct object type
+
+    :param document:
+    document -- The json
+    type_ -- The target object type
+    """
+    if isinstance(document, dict):
+        for k, v in document.items():
+            document[k] = _convert_object_type(v, type_)
+    if isinstance(document, list):
+        document = filter(_convert_object_type, type_)
+    elif isinstance(document, (str, bytes)):
+        try:
+            if type_ == 'ISODate':
+                return datetime.datetime.strptime(document, '%Y-%m-%dT%H:%M:%S.%fZ')
+            elif type_ == 'UUID':
+                return bson.Binary(uuid.UUID(document).bytes, 4)
```

Comment:
By default pymongo driver converts uuids to legacy 0x03 subtype.
This can be configured, but to be safe I added a check to make sure it was created with the correct subtype.